# Swetha Chandrasekar
# 012497628
# Exploring the use of word vectors and doc vectors as features for predicting fake news

# <u>Business problem </u>
<h1> Given a 'text' of statement made by a politician and 'the party affliation', predict whether the news is fake news or not </h1>
<h2> dataset: LIAR dataset </h2>
<h3>labelling <br>

<font color = red>0 - news labelled as [ false, barely true ] </font> <b> in LIAR dataset </b> <br>
<font color=green> 1 - new labelled as [true, mostly true, half true] <b> in LIAR dataset </b></font>
</h3>

In [13]:
lines = open('train.tsv','r',encoding='utf-8').readlines()

In [14]:
#Size of the dataset LIAR
len(lines)


10269

In [15]:
d =lines[0].strip().split('\t')
for i in range(0,len(d)):
    print (i,d[i])

0 2635.json
1 false
2 Says the Annies List political group supports third-trimester abortions on demand.
3 abortion
4 dwayne-bohac
5 State representative
6 Texas
7 republican
8 0
9 1
10 0
11 0
12 0
13 a mailer


<h1> Label 1 for true , Label 0 for fake news </h1>
<h2> true and mostly-true news are marked as 1 and false, barely-true and half-true are marked as 0 </h2>

In [16]:
true_data = []
false_data = []
overall_data = []
label = []
for i in range(0,len(lines)):
    line = lines[i].strip().lower().split()
    overall_data.append(lines[i].strip().lower())
    if (line[1]== 'true' or line[1] =='mostly-true'):
        true_data.append(lines[i].strip().lower())
        label.append(1)
    else:
        false_data.append(lines[i].strip().lower())
        label.append(0)
## Please note overall data still contains all of training data

# Data preprocessing
1. word tokenization
2. stop words removal
3. punctuation removal


<h1>use nltk word tokenize and remove stop words </h1>



In [17]:
##use nltk word tokenize and remove stop words
import nltk;
from nltk import word_tokenize
from nltk.corpus import stopwords
import string
stop = stopwords.words('english') + list(string.punctuation)
data = []
for line in overall_data:
    text = line.split('\t')[2].lower()
    party = line.split('\t')[7].lower()
    line = party + ' ' + text 
    tokens = [i for i in word_tokenize(line.strip().lower()) if i not in stop]
    bigrams = []
    for i in range(0,len(tokens)-1):
        bigram = tokens[i]+'-'+tokens[i+1]
        bigrams.append(bigram)
    for bigram in bigrams:
        tokens.append(bigram)
        
    data.append(tokens)
data[0]

['republican',
 'says',
 'annies',
 'list',
 'political',
 'group',
 'supports',
 'third-trimester',
 'abortions',
 'demand',
 'republican-says',
 'says-annies',
 'annies-list',
 'list-political',
 'political-group',
 'group-supports',
 'supports-third-trimester',
 'third-trimester-abortions',
 'abortions-demand']

<h1> Use Gensim to create word vectors from the dataset </h1>

In [18]:
# code referenced from https://machinelearningmastery.com/develop-word-embeddings-python-gensim/
from gensim.models import Word2Vec
model = Word2Vec(data, min_count=1)
# summarize the loaded model
print(model)
# summarize vocabulary
words = list(model.wv.vocab)
#print(words)
# access vector for one word
print(model['democrat'])
# save model
model.save('overall_model.bin')
# load model
overall_model = Word2Vec.load('overall_model.bin')
print(overall_model)
word_vector = overall_model

Word2Vec(vocab=92977, size=100, alpha=0.025)
[ 0.31897926 -0.30075017  0.44385406 -0.17423193  0.1243563  -0.33319739
  0.2366526   0.13807273 -0.06000553  0.15087457 -0.18188144 -0.51663405
  0.23263232  0.08559159  0.10304885 -0.07093515 -0.4035804   0.7325571
 -0.02796763  0.17175746  0.06813262  0.20064272  0.23084304 -0.0850437
  0.14919935 -0.31115928  0.21708335 -0.18335776 -0.45425603  0.10432383
 -0.47049436  0.0792982  -0.12893318 -0.20049219  0.01028899  0.14515536
  0.01766144  0.6455017  -0.9117627   0.0977454  -0.0754137  -0.23475714
  0.14027558  0.17047912  0.42759764 -0.11709759 -0.0266041  -0.24688077
  0.37782544  0.20485552 -0.17458831  0.04581442  0.3875592  -0.10467814
  0.11258423 -0.07504772  0.11266632 -0.17260388 -0.01944898  0.33609337
 -0.02069018  0.40547213 -0.27306816 -0.00265422  0.19421665 -0.03756718
 -0.19921534 -0.31526777 -0.22197212  0.6758196   0.5042538   0.09248549
  0.04989912  0.04865578 -0.19948936  0.37930933 -0.07374278  0.5284764
  0.32275

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:10: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  # Remove the CWD from sys.path while we load stuff.


Word2Vec(vocab=92977, size=100, alpha=0.025)


In [19]:
word_vector["obama"]

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  """Entry point for launching an IPython kernel.


array([ 0.33897433, -0.59048635,  0.6557907 , -0.33503222,  0.54515976,
       -0.47942203,  0.4455034 , -0.08991055,  0.22880368,  0.19982775,
        0.41465282, -0.6330831 ,  0.25364527,  0.3487599 ,  0.48313352,
       -0.2593625 , -0.34602714,  0.97673786, -0.31028423, -0.60180557,
        0.4787836 ,  0.5255844 ,  0.6542921 , -0.14999017,  0.37448248,
       -0.53197604,  0.8901366 ,  0.12900764, -0.7812516 ,  0.05211644,
       -0.31702915,  0.43272883, -0.35417983, -0.5982939 ,  0.04467162,
        0.22235258, -0.48438385,  1.0986974 , -1.0969946 , -0.11615343,
        0.5427212 , -0.02426909, -0.26657188,  0.3559324 ,  0.51013803,
       -0.34641662, -0.5705519 , -0.6110502 ,  0.05419513,  0.28984833,
       -0.14748605,  0.00824218,  0.39754778, -0.1230918 ,  0.1646678 ,
        0.17900738, -0.08766048, -0.06761001, -0.329739  ,  0.11669856,
        0.18208218,  0.17740816, -0.22810547, -0.55386263,  0.34158117,
        0.51130086, -0.27507374, -0.6103686 , -0.8581103 ,  0.70

## Word vectors represent latent signals projecting word as a vector in an n-dimensional space
## Let us use word vectors directly as features to predict if the news is fake or not
word vector simply gives array of numbers of size n. <br> <b>
<font color=green>We can directly use the sum of word vectors as features.</font> <br> </b>
## 1. sentences are made of multiple words, 
## 2. then we simply add the vectors of each word and get resultant vector for that sentence and use that as feature
## example vector('obama is socialist') = vector['obama'] + vector['is']+  vector['socialist']
## <font color=red> This may not be the right approach as doc2vec is a better approach. This was suggested by professor in class to compare it with doc2vec. </font>

In [20]:
X = []
for line in data:
    vectors = []
    for word in line:
        if word in overall_model:
            vector = overall_model[word]
            vectors.append(vector)
    word_vector = []

    for i in range(0,len(vectors[0])):
        val = 0.0
        for j in range(0,len(vectors)):
            val += vectors[j][i]
        word_vector.append(val)
    X.append(word_vector)
        


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:5: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  """
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:6: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  


# Run Logistic regression on features from word vector to label (0,1)

In [25]:
import pandas as pd
X_df = pd.DataFrame(X)
Y = pd.DataFrame(label)
scaled_X = X_df
scaled_Y = Y
from sklearn.model_selection import train_test_split
from sklearn import linear_model
X_train, X_test, y_train, y_test = train_test_split(scaled_X, scaled_Y, test_size=0.25, random_state=0)
lm = linear_model.LogisticRegression(verbose=1)
model = lm.fit(X_train, y_train)
print (model)
predictions = lm.predict(X_test)

print ("Score:", model.score(X_test, y_test))
w2vecmodel = model

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


[LibLinear]LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=1, warm_start=False)
Score: 0.6425233644859814


## Logistic regression was trained on word vectors and it has accuracy of 0.64 in predicting the fake news on test set. let us test on sample sentence 
lets say if republican says "obama supports third trimester abortion"

In [26]:
testing= "republican says obama supports third trimester abortion"

import numpy as np
def getVector(s):
    s = s.lower()
    words = word_tokenize(s)
    vectors = []
    for word in words:
        if word in overall_model:
            vectors.append(overall_model[word])
    sum_vector = []
    for i in range(0,len(vectors[0])):
        val = 0.0
        for j in range(0,len(vectors)):
            val += vectors[j][i]
        sum_vector.append(val)
    return pd.DataFrame([sum_vector])

testX = getVector(testing)
result_score = model.predict(testX)
model.predict_proba(testX)
### 0.71 probability that news belong to class 0 which is fake news

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:9: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  if __name__ == '__main__':
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:10: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  # Remove the CWD from sys.path while we load stuff.


array([[0.74095653, 0.25904347]])

## Class 1 probability is 0.28 and class 0 probability is 0.72 , so its likely a <font color=red> fake news </font> that <font color=red> "republican says obama supports third trimester abortion" </font>

# So far, we tried word2vec and summed the word vectors of individual words to get vector for each text

# Let's try doc2vec to see if embeddings for text as a whole is performs better than word2vec

In [27]:
#https://medium.com/@mishra.thedeepak/doc2vec-simple-implementation-example-df2afbbfbad5
from gensim.models.doc2vec import Doc2Vec, TaggedDocument
from nltk.tokenize import word_tokenize
data = []
for line in overall_data:
    T = line.strip().split('\t')
    text = T[2]
    party = T[7]
    text = party + ' ' + text
    data.append(text.lower())
tagged_data = [TaggedDocument(words=word_tokenize(_d.lower()), tags=[str(i)]) for i, _d in enumerate(data)]
data[0]

'republican says the annies list political group supports third-trimester abortions on demand.'

In [28]:
max_epochs = 10
vec_size = 20
alpha = 0.025

model = Doc2Vec(size=vec_size,
                alpha=alpha, 
                min_alpha=0.00025,
                min_count=1,
                dm =1)
  
model.build_vocab(tagged_data)

for epoch in range(max_epochs):
    print('iteration {0}'.format(epoch))
    model.train(tagged_data,
                total_examples=model.corpus_count,
                epochs=model.iter)
    # decrease the learning rate
    model.alpha -= 0.0002
    # fix the learning rate, no decay
    model.min_alpha = model.alpha

model.save("d2v.model")
print("Model Saved")

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/gensim/models/doc2vec.py:570: UserWarning: The parameter `size` is deprecated, will be removed in 4.0.0, use `vector_size` instead.
  warnings.warn("The parameter `size` is deprecated, will be removed in 4.0.0, use `vector_size` instead.")


iteration 0


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:17: DeprecationWarning: Call to deprecated `iter` (Attribute will be removed in 4.0.0, use self.epochs instead).


iteration 1
iteration 2
iteration 3
iteration 4
iteration 5
iteration 6
iteration 7
iteration 8
iteration 9
Model Saved


In [29]:
from gensim.models.doc2vec import Doc2Vec

model= Doc2Vec.load("d2v.model")
#to find the vector of a document which is not in training data
test_data = word_tokenize("republicans say obama supports third trimester abortion".lower())
v1 = model.infer_vector(test_data)
v1


array([-0.0183533 ,  0.01229155, -0.00439514,  0.17856328,  0.06666941,
       -0.0116841 ,  0.1876027 ,  0.28720534, -0.12316225,  0.0389136 ,
       -0.02605944, -0.11202887,  0.14041743,  0.11761943, -0.1597844 ,
        0.1690564 ,  0.16085854,  0.01233068,  0.10842273,  0.08658281],
      dtype=float32)

In [30]:
doc2vecFeatures = []
for line in overall_data:
    T = line.strip().split('\t')
    text = T[2]
    party = T[7]
    text = party + ' ' + text
    doc2vecFeatures.append(model.infer_vector(word_tokenize(text)))

# Let us use doc2vec vector embeddings directly as a feature and measure its performance against word2vec vector sums

In [31]:
import pandas as pd
X_df = pd.DataFrame(doc2vecFeatures)
Y = pd.DataFrame(label)
scaled_X = X_df
scaled_Y = Y
from sklearn.model_selection import train_test_split
from sklearn import linear_model
X_train, X_test, y_train, y_test = train_test_split(scaled_X, scaled_Y, test_size=0.3, random_state=0)
lm = linear_model.LogisticRegression(verbose=1)
model = lm.fit(X_train, y_train)
print (model)
predictions = lm.predict(X_test)

print ("Score:", model.score(X_test, y_test))

[LibLinear]LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=1, warm_start=False)
Score: 0.6416747809152873


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/sklearn/utils/validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [32]:
text = "republican says obama supports third trimester abortion"
dvmodel= Doc2Vec.load("d2v.model")
test_sentence= [dvmodel.infer_vector(word_tokenize(text))]

In [33]:
predicted = model.predict(pd.DataFrame(test_sentence))
predicted_probabilities = model.predict_proba(pd.DataFrame(test_sentence))

In [34]:
print ("Doc2Vec prediction")
print ("=====================")
print ("Predicted class:",predicted, "Predicted probabilities for class[0,1]:",predicted_probabilities)
print ("Length of embeddings from Doc2Vec:", len(dvmodel.infer_vector(word_tokenize(text))))
df = getVector(text)
print ("Length of embeddings from Word2Vec embeddings:",len(df.columns))
print ("====================================================================")


Doc2Vec prediction
Predicted class: [0] Predicted probabilities for class[0,1]: [[0.6112897 0.3887103]]
Length of embeddings from Doc2Vec: 20
Length of embeddings from Word2Vec embeddings: 100


/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:9: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  if __name__ == '__main__':
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:10: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  # Remove the CWD from sys.path while we load stuff.


## Class 1 probability is 0.35 and class 0 probability is 0.64 , so its likely a <font color=red> fake news </font> that <font color=red> "republican says obama supports third trimester abortion" </font>

# Accuracy in predicting fake news based on 
## Doc2Vec embeddings: 0.641
##  Word2Vec embeddings: 0.643

# word2vec embeddings with vector summation seems to perform similar to Doc2Vec embeddings
# it could be because word2vec is generating vectors of size <u>100</u> and while doc2vec is generating vectors of size <u>20.</u> 
# Doc 2 vec seem to get the same accuracy with 20 embeddings that word2vec does with 100 embeddings.